coordinator/src/main/scala/filodb.coordinator/IngestionActor.scala

```diff
@@ -82,66 +83,105 @@ private[filodb] final class IngestionActor(dataset: Dataset,
+    * reconciles any differences. It does so by stopping ingestion for shards that aren't mapped
+    * to this node, and it starts ingestion for those that are.
     */
-  private def start(e: StartShardIngestion, origin: ActorRef): Unit =
-    if (invalid(e.ref)) handleInvalid(e, Some(origin)) else {
-      try memStore.setup(dataset, e.shard, storeConfig, downsample) catch {
-        case ShardAlreadySetup(ds, shard) =>
-          logger.warn(s"dataset=$ds shard=$shard already setup, skipping....")
-          return
-      }
+  private def resync(state: ShardIngestionState, origin: ActorRef): Unit = {
+    if (invalid(state.ref)) {
+      logger.error(s"$state is invalid for this ingester '${dataset.ref}'.")
+      return
+    }
 
-      val ingestion = for {
-        _ <- memStore.recoverIndex(dataset.ref, e.shard)
-        checkpoints <- memStore.metastore.readCheckpoints(dataset.ref, e.shard) }
-      yield {
-        if (checkpoints.isEmpty) {
-          // Start normal ingestion with no recovery checkpoint and flush group 0 first
-          normalIngestion(e.shard, None, 0, storeConfig.diskTTLSeconds)
-        } else {
-          // Figure out recovery end watermark and intervals.  The reportingInterval is the interval at which
-          // offsets come back from the MemStore for us to report progress.
-          val startRecoveryWatermark = checkpoints.values.min + 1
-          val endRecoveryWatermark = checkpoints.values.max
-          val lastFlushedGroup = checkpoints.find(_._2 == endRecoveryWatermark).get._1
-          val reportingInterval = Math.max((endRecoveryWatermark - startRecoveryWatermark) / 20, 1L)
-          logger.info(s"Starting recovery for dataset=${dataset.ref} " +
-            s"shard=${e.shard}: from $startRecoveryWatermark to $endRecoveryWatermark; " +
-            s"last flushed group $lastFlushedGroup")
-          logger.info(s"Checkpoints for dataset=${dataset.ref} shard=${e.shard}: $checkpoints")
-          for { lastOffset <- doRecovery(e.shard, startRecoveryWatermark, endRecoveryWatermark, reportingInterval,
-                                         checkpoints) }
-          yield {
-            // Start reading past last offset for normal records; start flushes one group past last group
-            normalIngestion(e.shard, Some(lastOffset + 1), (lastFlushedGroup + 1) % numGroups,
-                            storeConfig.diskTTLSeconds)
+    if (state.version != 0 && state.version <= shardStateVersion) {
+      logger.info(s"Ignoring old ShardIngestionState version: ${state.version} <= $shardStateVersion")
+      return
+    }
+
+    // Start with the full set of all shards being ingested, and remove shards from this set
+    // which must continue being ingested.
+    val shardsToStop = HashSet() ++ streams.keySet
+
+    for (shard <- 0 until state.map.numShards) {
+      if (state.map.coordForShard(shard) == context.parent) {
```

Comment:
Does the == work reliably for local and remote actor comparison?

Reply:
I believe this was the same comparison as before. Equality checking for ActorRef considers the path and uid (according to the docs), so this comparison should work. Unlike Java ==, Scala calls the equals method.